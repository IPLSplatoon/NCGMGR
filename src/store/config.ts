import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const configStore = createStore<Configuration>({
    state: {
        installPath: ''
    },
    mutations: {
        setInstallPath: (state, installPath: string) => {
            state.installPath = installPath
        },
        setState: (state, config: Configuration) => {
            Object.assign(state, config)
        }
    },
    actions: {
        load: store => {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            store.commit('setState', {
                ...store.state,
                ...config
            })
        },
        persist: store => {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(store.state))
        }
    }
})

export const configStoreKey: InjectionKey<Store<Configuration>> = Symbol()

export function useConfigStore (): Store<Configuration> {
    return useStore(configStoreKey)
}
