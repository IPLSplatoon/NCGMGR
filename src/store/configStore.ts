import { defineStore } from 'pinia'

export const LOCAL_STORAGE_CONFIG_KEY = 'config'

export interface Configuration {
    installPath: string
}

export const useConfigStore = defineStore('config', {
    state: () => ({
        installPath: ''
    } as Configuration),
    actions: {
        load () {
            const config = JSON.parse(localStorage.getItem(LOCAL_STORAGE_CONFIG_KEY) ?? '{}')
            this.$patch(config)
        },
        persist () {
            localStorage.setItem(LOCAL_STORAGE_CONFIG_KEY, JSON.stringify(this.$state))
        }
    }
})
