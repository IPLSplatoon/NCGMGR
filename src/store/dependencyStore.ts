import { defineStore } from 'pinia'
import { isBlank } from '@iplsplatoon/vue-components'
import { getNodejsVersion } from '@/service/dependencyService'

interface DependencyStore {
    isLoading: boolean
    nodejs: {
        version: string | null
    }
}

export const useDependencyStore = defineStore('dependency', {
    state: () => ({
        isLoading: false,
        nodejs: {
            version: null
        }
    } as DependencyStore),
    getters: {
        hasNodejs: state => !isBlank(state.nodejs.version)
    },
    actions: {
        async checkVersions () {
            this.isLoading = true
            try {
                this.nodejs.version = await getNodejsVersion()
            } finally {
                this.isLoading = false
            }
        }
    }
})
