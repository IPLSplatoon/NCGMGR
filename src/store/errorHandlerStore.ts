import { App } from 'vue'
import { defineStore } from 'pinia'
import { generateId } from '@/util/generateId'

interface ErrorHandlerStore {
    recentErrors: Record<string, unknown>
}

export const useErrorHandlerStore = defineStore('errorHandler', {
    state: () => ({
        recentErrors: {}
    } as ErrorHandlerStore),
    actions: {
        removeRecentError ({ key }: { key: string }): void {
            delete this.recentErrors[key]
        },
        handleError ({ err, info }: { err: unknown, info: string }): void {
            console.error(`Got error from '${info}': \n`, err)

            const id = generateId()
            this.recentErrors[id] = err
        }
    }
})

export function setUpErrorHandler (app: App<unknown>): void {
    const store = useErrorHandlerStore()
    app.config.errorHandler = (err, vm, info) => {
        store.handleError({ err, info })
    }
}
