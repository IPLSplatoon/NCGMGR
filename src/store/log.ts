import { LogEvent } from '@/types/log'
import { createStore, Store, useStore } from 'vuex'
import { InjectionKey } from 'vue'
import { Event, listen, UnlistenFn } from '@tauri-apps/api/event'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    completed: Record<string, boolean>
}

const unlistenFns: Record<string, UnlistenFn> = {}

export const logStore = createStore<LogStore>({
    state: {
        lines: {},
        completed: {}
    },
    mutations: {
        insertLine (state, { line, key }: { line: LogEvent, key: string }) {
            if (!state.lines[key]) {
                state.lines[key] = []
            }
            state.lines[key].push(line)
        },
        reset (state, key: string) {
            state.lines[key] = []
            state.completed[key] = false
        },
        setCompleted (state, { key, completed }: { key: string, completed: boolean }) {
            state.completed[key] = completed
        }
    },
    actions: {
        listen (store, key: string) {
            if (unlistenFns[key]) {
                return
            }
            return listen(`log:${key}`, (event: Event<LogEvent>) => {
                store.commit('insertLine', { line: event.payload, key })
            }).then(unlisten => {
                unlistenFns[key] = unlisten
            })
        },
        unlisten (store, key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key]()
            }
        },
        logPromiseResult (store, { promise, key, noLogOnSuccess }: { promise: Promise<unknown>, key: string, noLogOnSuccess: boolean }) {
            promise.then(() => {
                if (!noLogOnSuccess) {
                    store.commit('insertLine', {
                        line: {
                            message: 'Success!',
                            type: 'success'
                        },
                        key
                    })
                }
            }).catch(e => {
                store.commit('insertLine', {
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
            }).finally(() => {
                store.commit('setCompleted', { key, completed: true })
            })
        }
    }
})

export const logStoreKey: InjectionKey<Store<LogStore>> = Symbol()

export function useLogStore (): Store<LogStore> {
    return useStore(logStoreKey)
}
