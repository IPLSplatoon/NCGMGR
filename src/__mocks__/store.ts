import { createStore, Store } from 'vuex'
import { Configuration } from '@/store/config'
import { LogStore } from '@/store/log'

export function createConfigStore (): Store<Configuration> {
    return createStore<Configuration>({
        state: {
            installPath: '/install/path'
        },
        actions: {
            load: jest.fn(),
            persist: jest.fn()
        },
        mutations: {
            setInstallPath: jest.fn()
        }
    })
}

export function createLogStore (): Store<LogStore> {
    return createStore<LogStore>({
        state: {
            lines: [],
            completed: false
        },
        mutations: {
            reset: jest.fn(),
            insertLine: jest.fn(),
            setCompleted: jest.fn()
        }
    })
}
