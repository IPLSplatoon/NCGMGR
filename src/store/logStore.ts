import { ActionState, LogEvent, ProgressEvent } from '@/types/log'
import { listen, UnlistenFn } from '@tauri-apps/api/event'
import { defineStore } from 'pinia'

export interface LogStore {
    lines: Record<string, LogEvent[]>
    progressEntries: Record<string, ProgressEvent>
    actionStates: Record<string, ActionState>
}

const unlistenFns: Record<string, UnlistenFn[]> = {}

export const useLogStore = defineStore('log', {
    state: () => ({
        lines: {},
        progressEntries: {},
        actionStates: {}
    } as LogStore),
    actions: {
        insertLine ({ line, key }: { line: LogEvent, key: string }) {
            if (!this.lines[key]) {
                this.lines[key] = []
            }
            this.lines[key].push(line)
        },
        setProgress (key: string, event: ProgressEvent) {
            const existingEntry = this.progressEntries[key]
            if (existingEntry == null) {
                this.progressEntries[key] = event
                this.logProgressEvent(key, event)
            } else {
                existingEntry.message = event.message
                if (event.maxStep != null) {
                    existingEntry.maxStep = event.maxStep
                }
                if (event.step != null) {
                    existingEntry.step = event.step

                    if (event.step === event.maxStep) {
                        this.setActionState({ key, state: ActionState.COMPLETED_SUCCESS })
                    }
                }

                this.logProgressEvent(key, {
                    message: event.message,
                    step: event.step,
                    maxStep: existingEntry.maxStep
                })
            }
        },
        logProgressEvent(key: string, event: ProgressEvent) {
            if (event.step != null) {
                this.insertLine({
                    line: {
                        message: `[${event.step}/${event.maxStep}] ${event.message}`,
                        type: 'info'
                    },
                    key
                })
            } else {
                this.insertLine({
                    line: {
                        message: event.message,
                        type: 'info'
                    },
                    key
                })
            }
        },
        reset (key: string) {
            delete this.progressEntries[key]
            this.lines[key] = []
            this.actionStates[key] = ActionState.INCOMPLETE
        },
        setActionState ({ key, state }: { key: string, state: ActionState }) {
            this.actionStates[key] = state
        },
        listen (key: string, listenForProgress = false) {
            if (unlistenFns[key]) {
                return
            }

            const listenPromises: Array<Promise<UnlistenFn>> = []

            listenPromises.push(listen<LogEvent>(`log:${key}`, event => {
                this.insertLine({ line: event.payload, key })
            }))

            if (listenForProgress) {
                listenPromises.push(listen<ProgressEvent>(`progress:${key}`, event => {
                    this.setProgress(key, event.payload)
                }))
            }

            return Promise.all(listenPromises).then(result => {
                unlistenFns[key] = result
            })
        },
        unlisten (key: string) {
            if (unlistenFns[key]) {
                unlistenFns[key].forEach(fn => fn())
                delete unlistenFns[key]
            }
        },
        logPromiseResult ({ promise, key }: { promise: Promise<unknown>, key: string }) {
            promise.catch(e => {
                this.insertLine({
                    line: {
                        message: String(e),
                        type: 'error'
                    },
                    key
                })
                this.setActionState({ key, state: ActionState.COMPLETED_ERROR })
            })
        }
    }
})
