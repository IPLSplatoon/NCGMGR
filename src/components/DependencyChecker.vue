<template>
    <ipl-expanding-space title="Dependency check" :expanded="!dependencyStore.hasNodejs">
        <status-row :color="nodejsStatusColor" class="m-t-6">
            {{ nodejsStatusText }}
        </status-row>
        <ipl-button class="m-t-8" label="Check again" @click="runDependencyCheck" />
    </ipl-expanding-space>
</template>

<script lang="ts">
import { defineComponent } from '@vue/runtime-core'
import { IplButton, IplExpandingSpace } from '@iplsplatoon/vue-components'
import { useDependencyStore } from '@/store/dependencyStore'
import { computed, onMounted } from 'vue'
import StatusRow from '@/components/statusRow.vue'

export default defineComponent({
    name: 'DependencyChecker',

    components: { IplButton, StatusRow, IplExpandingSpace },

    setup () {
        const dependencyStore = useDependencyStore()

        onMounted(() => {
            dependencyStore.checkVersions()
        })

        return {
            dependencyStore,
            nodejsStatusColor: computed(() => {
                if (dependencyStore.isLoading) return 'yellow'
                else if (dependencyStore.hasNodejs) return 'green'
                else return 'red'
            }),
            nodejsStatusText: computed(() => {
                if (dependencyStore.isLoading) return 'Loading...'
                else if (dependencyStore.hasNodejs) return `Found Node.js ${dependencyStore.nodejs.version}.`
                else return 'Node.js not found.'
            }),
            runDependencyCheck () {
                dependencyStore.checkVersions()
            }
        }
    }
})
</script>
