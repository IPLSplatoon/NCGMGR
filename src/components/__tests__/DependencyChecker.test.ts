import DependencyChecker from '@/components/DependencyChecker.vue'
import { config, mount } from '@vue/test-utils'
import { createTestingPinia, TestingPinia } from '@pinia/testing'
import { useDependencyStore } from '@/store/dependencyStore'

describe('DependencyChecker', () => {
    config.global.stubs = {
        IplButton: true,
        FontAwesomeIcon: true,
        StatusRow: true
    }

    let pinia: TestingPinia

    beforeEach(() => {
        pinia = createTestingPinia()
        config.global.plugins = [pinia]
    })

    it('matches snapshot when dependencies are loading', () => {
        useDependencyStore().isLoading = true

        expect(mount(DependencyChecker).html()).toMatchSnapshot()
    })

    it('matches snapshot when node.js is missing', () => {
        const dependencyStore = useDependencyStore()
        dependencyStore.isLoading = false
        dependencyStore.nodejs.version = null

        expect(mount(DependencyChecker).html()).toMatchSnapshot()
    })

    it('matches snapshot when dependencies are present', () => {
        const dependencyStore = useDependencyStore()
        dependencyStore.isLoading = false
        dependencyStore.nodejs.version = '12.3.4'

        expect(mount(DependencyChecker).html()).toMatchSnapshot()
    })
})
