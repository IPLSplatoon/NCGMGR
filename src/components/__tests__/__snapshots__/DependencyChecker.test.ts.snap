// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DependencyChecker matches snapshot when dependencies are loading 1`] = `
<div class="ipl-expansion-panel__content" data-v-1dcbcc2a="">
  <div class="ipl-expansion-panel__header layout horizontal center-vertical" data-v-1dcbcc2a="">
    <div class="ipl-expansion-panel__header-background" data-v-1dcbcc2a=""></div>
    <div class="ipl-expansion-panel__title" data-v-1dcbcc2a="">Dependency check</div>
    <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon content-expanded" data-v-1dcbcc2a=""></font-awesome-icon-stub>
    <div class="header-extra" data-v-1dcbcc2a=""></div>
  </div>
  <div class="content" data-v-1dcbcc2a="">
    <div class="status-row layout horizontal center-vertical m-t-6">
      <div class="status-row__status-light color-yellow"></div>Loading...
    </div>
    <ipl-button-stub label="Check again" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" class="m-t-8"></ipl-button-stub>
  </div>
</div>
`;

exports[`DependencyChecker matches snapshot when dependencies are present 1`] = `
<div class="ipl-expansion-panel__content" data-v-1dcbcc2a="">
  <div class="ipl-expansion-panel__header layout horizontal center-vertical" data-v-1dcbcc2a="">
    <div class="ipl-expansion-panel__header-background" data-v-1dcbcc2a=""></div>
    <div class="ipl-expansion-panel__title" data-v-1dcbcc2a="">Dependency check</div>
    <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon" data-v-1dcbcc2a=""></font-awesome-icon-stub>
    <div class="header-extra" data-v-1dcbcc2a=""></div>
  </div>
  <div class="content" data-v-1dcbcc2a="" style="display: none;">
    <div class="status-row layout horizontal center-vertical m-t-6">
      <div class="status-row__status-light color-green"></div>Found Node.js 12.3.4.
    </div>
    <ipl-button-stub label="Check again" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" class="m-t-8"></ipl-button-stub>
  </div>
</div>
`;

exports[`DependencyChecker matches snapshot when node.js is missing 1`] = `
<div class="ipl-expansion-panel__content" data-v-1dcbcc2a="">
  <div class="ipl-expansion-panel__header layout horizontal center-vertical" data-v-1dcbcc2a="">
    <div class="ipl-expansion-panel__header-background" data-v-1dcbcc2a=""></div>
    <div class="ipl-expansion-panel__title" data-v-1dcbcc2a="">Dependency check</div>
    <font-awesome-icon-stub icon="chevron-left" border="false" fixedwidth="false" listitem="false" pulse="false" swapopacity="false" spin="false" symbol="false" inverse="false" class="icon content-expanded" data-v-1dcbcc2a=""></font-awesome-icon-stub>
    <div class="header-extra" data-v-1dcbcc2a=""></div>
  </div>
  <div class="content" data-v-1dcbcc2a="">
    <div class="status-row layout horizontal center-vertical m-t-6">
      <div class="status-row__status-light color-red"></div>Node.js not found.
    </div>
    <ipl-button-stub label="Check again" color="blue" disabled="false" async="false" progressmessage="Working..." successmessage="Done!" requiresconfirmation="false" shortconfirmationmessage="false" small="false" disableonsuccess="false" href="" class="m-t-8"></ipl-button-stub>
  </div>
</div>
`;
